File: backend/app/schemas/scraper.py
"""
Scraper Schemas
Pydantic models for scraper-related operations
"""

from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime


class ScraperRequest(BaseModel):
    """Request to trigger scraping"""
    sites: Optional[List[str]] = Field(default=None, description="Specific sites to scrape (if None, scrapes all user's enabled sites)")
    save_to_db: bool = Field(default=True, description="Save scraped articles to database")


class ArticleResponse(BaseModel):
    """Single article response"""
    id: Optional[int] = None
    source: str
    publisher: Optional[str] = None
    headline: str
    article_url: str
    published_time: Optional[str] = None
    country: Optional[str] = None
    view: Optional[str] = None
    extra_data: Optional[Dict[str, Any]] = None
    scraped_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class ScraperStatus(BaseModel):
    """Scraper job status"""
    job_id: int
    status: str  # pending, running, completed, failed
    progress: int  # 0-100
    status_message: Optional[str] = None
    articles_count: Optional[int] = None
    current_site: Optional[str] = None
    sites_completed: Optional[int] = None
    total_sites: Optional[int] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    error: Optional[str] = None


class ScraperResult(BaseModel):
    """Scraper job result"""
    job_id: int
    status: str
    total_articles: int
    articles_by_site: Dict[str, int]
    articles: List[ArticleResponse]
    completed_at: Optional[datetime] = None


class SiteConfig(BaseModel):
    """Site configuration response"""
    name: str
    url: str
    enabled: bool
    description: Optional[str] = None


class UserSitesResponse(BaseModel):
    """User's available scraper sites"""
    enabled_sites: List[str]
    available_sites: List[SiteConfig]
    default_sites: List[str] = []
    use_custom_default: bool = False
    allowed_sites: List[str] = []  # Admin-restricted sites (empty = all)


class UpdateSitesRequest(BaseModel):
    """Request to update user's enabled sites"""
    enabled_sites: List[str] = Field(..., description="List of site names to enable")


class SetDefaultSitesRequest(BaseModel):
    """Request to set current sites as default"""
    set_as_default: bool = Field(default=True, description="Set current enabled sites as default")


class SitesUpdateResponse(BaseModel):
    """Response after updating sites"""
    success: bool
    enabled_sites: List[str]
    default_sites: List[str]
    use_custom_default: bool
    message: str


# Admin schemas for site assignment
class AdminAssignSitesRequest(BaseModel):
    """Admin request to assign allowed sites to a user"""
    allowed_sites: List[str] = Field(
        default=[],
        description="List of site names user can access. Empty list = access to ALL sites."
    )


class AdminUserSitesResponse(BaseModel):
    """Admin response for user's site configuration"""
    user_id: int
    username: str
    email: str
    allowed_sites: List[str]
    enabled_sites: List[str]
    all_available_sites: List[str]
    message: str = ""

File: frontend/src/api/scraper.ts
import axios from './axios';

export interface ScraperSite {
  id?: number;
  name: string;
  url: string;
  enabled: boolean;
  description?: string;
}

export interface UserSitesResponse {
  enabled_sites: string[];
  available_sites: ScraperSite[];
  default_sites: string[];
  use_custom_default: boolean;
}

export interface SitesUpdateResponse {
  success: boolean;
  enabled_sites: string[];
  default_sites: string[];
  use_custom_default: boolean;
  message: string;
}

export interface ScraperJob {
  job_id: string;
  status: 'pending' | 'running' | 'completed' | 'failed';
  progress?: number;
  current_site?: string;
  articles_scraped?: number;
  sites_completed?: number;
  total_sites?: number;
  error?: string;
  created_at: string;
  completed_at?: string;
}

export interface ScraperResult {
  total_articles: number;
  sites_scraped: number;
  new_articles: number;
  articles: any[];
}

// Start scraper job
export const startScraper = async (
  sites?: string[],
  views?: string[]
): Promise<ScraperJob> => {
  const response = await axios.post('/api/scraper/run', {
    sites,
    views,
  });
  return response.data;
};

// Get scraper job status
export const getScraperStatus = async (jobId: string): Promise<ScraperJob> => {
  const response = await axios.get(`/api/scraper/status/${jobId}`);
  return response.data;
};

// Get scraper job result
export const getScraperResult = async (jobId: string): Promise<ScraperResult> => {
  const response = await axios.get(`/api/scraper/result/${jobId}`);
  return response.data;
};

// Get available sites with user's enabled status
export const getScraperSites = async (): Promise<UserSitesResponse> => {
  const response = await axios.get('/api/scraper/sites');
  return response.data;
};

// Update user's enabled sites (instant save on toggle)
export const updateEnabledSites = async (enabledSites: string[]): Promise<SitesUpdateResponse> => {
  const response = await axios.put('/api/scraper/sites', {
    enabled_sites: enabledSites,
  });
  return response.data;
};

// Set current enabled sites as user's default
export const setDefaultSites = async (): Promise<SitesUpdateResponse> => {
  const response = await axios.post('/api/scraper/sites/default');
  return response.data;
};

// Clear custom default and use system default
export const clearDefaultSites = async (): Promise<SitesUpdateResponse> => {
  const response = await axios.delete('/api/scraper/sites/default');
  return response.data;
};

// Get scraper job history
export const getScraperJobs = async (
  page: number = 1,
  size: number = 10
): Promise<{
  items: ScraperJob[];
  total: number;
  page: number;
  size: number;
  pages: number;
}> => {
  const response = await axios.get('/api/scraper/jobs', {
    params: { skip: (page - 1) * size, limit: size },
  });
  return response.data;
};
